Added zoomable stacked(locations) bar chart with tooltips for offer events.

muthurajendran opened the pull request.

muthurajendran committed:
added bar chart code

muthurajendran committed:
Merge branch 'development-isi' of https://github.com/usc-isi-i2/digapp-ht

muthurajendran committed:
remove heaad

muthurajendran committed:
added graph component

muthurajendran committed:
added tooltip and color changes

muthurajendran committed:
Merge branch 'master' of https://github.com/NextCenturyCorporation/digapp-ht

muthurajendran committed:
fixed the merge and code errors

muthurajendran committed:
deleted test files

muthurajendran committed:
remove unwanted files

muthurajendran committed:
Merge branch 'master' of https://github.com/NextCenturyCorporation/digapp-ht

ThomasSchellenbergNextCentury commented:
Is the new bar chart also going to be in the email entity page as well?  Edit:  Also, will it be used for the revisions in the offer entity page?

ThomasSchellenbergNextCentury commented:
Assuming that the new bar chart will replace the old one, you can remove the old bar chart sections from the entity pages as well as the components used to create the bar chart (offerTimeline) data.  Additionally, if the offerTimeline queryTemplate is no longer needed, you can remove it.

muthurajendran commented:
Let me confirm. @saggu do we replace the old ones? also add to the email page as well?

muthurajendran committed:
added and replaced offer timelines

muthurajendran committed:
Merge branch 'master' of https://github.com/NextCenturyCorporation/digapp-ht

muthurajendran commented:
I've removed the old graph and updated with the new one. The new graph has been pushed to the email, phone and seller page. 
I think the offerTimeline query template is still been used in other pages, so not sure if needed to be removed.

ThomasSchellenbergNextCentury commented on app/elements/bar-zoomable-chart/bar-zoomable-chart.html:
```diff
@@ -0,0 +1,316 @@
+        
+        .axis {
+          font-size: 11px;
+          fill: var(--bar-zoomable-chart-axis-color, --secondary-text-color);
```
Is the --bar-zoomable-chart-axis-color variable defined anywhere?  If not, you can remove it.

ThomasSchellenbergNextCentury commented on app/elements/bar-zoomable-chart/bar-zoomable-chart.html:
```diff
@@ -0,0 +1,316 @@
+    Polymer({
+      is: 'bar-zoomable-chart',
+      properties: {
+        data: {
```
It would be better if, instead of having a single "data" property that contained a "locations" array and a "data" array, this component had two separate properties that took the "locations" (data.locations) and "data" (data.data).  This helps make the component more reusable.

ThomasSchellenbergNextCentury commented on app/elements/bar-zoomable-chart/bar-zoomable-chart.html:
```diff
@@ -0,0 +1,316 @@
+
+        if (this.data.data) {
+          this.data.data.map(function(record) {
+            record.date = parseDate(new Date(record.date)); // turn the date string into a date object
```
This component shouldn't modify the input data because it would cause issues if you're using the same data in other components/pages.  Instead, add a new property to the component (something like "barChartData") and store the processed input data in that property.

ThomasSchellenbergNextCentury commented on app/scripts/seller-transform.js:
```diff
@@ -89,6 +89,39 @@ var sellerTransform = (function(_, relatedEntityTransform, commonTransforms) {
+            var point = {}
+            point.date = record.key_as_string;
+
+            point.value = {}
```
It's helpful to use a name that's not "value" when able since "value" is a very generic term.  In this case, the object represents a mapping from city names to their count, so maybe something like "cityCounts"?

ThomasSchellenbergNextCentury commented on app/scripts/seller-transform.js:
```diff
@@ -89,6 +89,39 @@ var sellerTransform = (function(_, relatedEntityTransform, commonTransforms) {
+
+            point.value = {}
+            if (record.localities.buckets){
+                sum = 0;
```
The "sum" (line 100), "geoData" (line 102), and "city" (line 103) variables have all been created here as global variables.  Please use "var" to create them as local variables.

ThomasSchellenbergNextCentury commented on app/scripts/seller-transform.js:
```diff
@@ -89,6 +89,39 @@ var sellerTransform = (function(_, relatedEntityTransform, commonTransforms) {
+    function processLocationGraph(records){
+        var data = [];
+        _.each(records, function(record){
+            var point = {}
```
Missing semicolons on lines 95, 98, and 119.

ThomasSchellenbergNextCentury commented:
I know this is nitpicking, but since "bar chart" itself is a commonly used phrase, it would be preferable if the name of this new component didn't split up "bar chart" by adding another word in the middle.  "Zoomable-bar-chart" would be preferable.

ThomasSchellenbergNextCentury commented on app/scripts/seller-transform.js:
```diff
@@ -192,6 +225,16 @@ var sellerTransform = (function(_, relatedEntityTransform, commonTransforms) {
             }
             return newData;
+        },
+        locationTimeline: function(data){
```
Unit tests for this new transform should be added to app/test/client-config-seller-test.html.

ThomasSchellenbergNextCentury commented:
I know Pedro said you hadn't had a chance to get to it yet but the new component will need its own set of unit tests.  I'm not sure about your experience with our testing framework so I've taken the liberty to create a test stub for you (see below).  Obviously it's overly complicated to test the D3 elements themselves but you can still check that the component's properties are getting set correctly (including the processed data) and the helper functions like getItemKey work as expected.  Take a look at app/test/bar-chart-basic.html for some examples for our old bar chart and let us know if you have any other questions.  Thanks!

```
<!doctype html>
<!--
@license
Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
Code distributed by Google as part of the polymer project is also
subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
-->

<html>
<head>
  <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes">
  <title>bar-chart</title>

  <script src="../bower_components/webcomponentsjs/webcomponents.min.js"></script>
  <script src="../bower_components/web-component-tester/browser.js"></script>
  <script src="../bower_components/test-fixture/test-fixture-mocha.js"></script>
  <link rel="import" href="../bower_components/test-fixture/test-fixture.html">

  <!-- Step 1: import the element to test -->
  <link rel="import" href="../elements/bar-zoomable-chart/bar-zoomable-chart.html">

</head>
<body>

<test-fixture id="zoomable-bar-chart-fixture">
  <template>
    <bar-zoomable-chart>
    </bar-zoomable-chart>
  </template>
</test-fixture>

<test-fixture id="zoomable-bar-chart-no-data-fixture">
  <template>
    <bar-zoomable-chart>
    </bar-zoomable-chart>
  </template>
</test-fixture>

<script>
  /* globals suite, setup, fixture, test, expect, d3 */
  /* jshint -W030 */
  suite('<zoomable-bar-chart>', function() {
    var zoomableBarChart;

    setup(function() {
      zoomableBarChart = fixture('zoomable-bar-chart-fixture');
    });
  });

  suite('<zoomable-bar-chart> with no data', function() {
    var zoomableBarChart;

    setup(function() {
      zoomableBarChart = fixture('zoomable-bar-chart-no-data-fixture');
    });
  });
</script>

</body>
</html>
```

ThomasSchellenbergNextCentury commented on app/elements/bar-zoomable-chart/bar-zoomable-chart.html:
```diff
@@ -0,0 +1,316 @@
+            for (var key in record.value) {
+              var obj = {};
+              obj.name = key;
+              obj.y0 = y0;
```
The variables y0 and y1 could be given more descriptive names like perhaps yAxisStart and yAxisEnd.

muthurajendran committed:
recommended fixes from next century

muthurajendran committed:
Merge https://github.com/NextCenturyCorporation/digapp-ht

muthurajendran committed:
Merge https://github.com/NextCenturyCorporation/digapp-ht

muthurajendran committed:
added transform test case and changes to vars

muthurajendran committed:
Merge https://github.com/NextCenturyCorporation/digapp-ht

muthurajendran committed:
Merge https://github.com/NextCenturyCorporation/digapp-ht

muthurajendran committed:
added test cases

muthurajendran commented:
I've done all the fixes and test cases! Let me know if we need any more
